[mmp][mtouch] Better guard framework SDK checks

chamons opened the pull request:
- https://github.com/xamarin/xamarin-macios/issues/3725
- These frameworks "CoreAudioKit Metal MetalKit MetalPerformanceShaders CoreNFC DeviceCheck"
 were special cased, but that special case did do an SDK check.
- Create a helper method to share check

chamons commented:
No test because I couldn't think of a cute way to handle this without requiring running mmp and\or mtouch on older macOS versions.

I manually tested it on 10.11.

rolfbjarne commented:
> No test because I couldn't think of a cute way to handle this without requiring running mmp and\or mtouch on older macOS versions.

You need to run on older Xcode's, not macOS versions.

There are already some logic to get older Xcode's here: https://github.com/xamarin/xamarin-macios/blob/379a4fb2f5052fa4d29375bcff5751e6a8903302/tests/common/Configuration.cs#L207-L210, it should just be a matter of adding one for Xcode 8 (or whatever you need) and then use that Xcode in your test.

spouliot commented on tools/common/Assembly.cs:
```diff
@@ -328,6 +328,15 @@ public static LinkWithAttribute GetLinkWithAttribute (CustomAttribute attr)
 
+		void AddFramework (string file)
+		{
+			Framework framework;
```
minor: there's a nicer syntax for this now :)

`TryGetValue (file, out var framework)`

spouliot commented on tools/common/Assembly.cs:
```diff
@@ -328,6 +328,15 @@ public static LinkWithAttribute GetLinkWithAttribute (CustomAttribute attr)
+		{
+			Framework framework;
+			if (Driver.GetFrameworks (App).TryGetValue (file, out framework) && framework.Version > App.SdkVersion)
+				Driver.Log (3, "Not linking with the framework {0} (referenced by a module reference in {1}) because it was introduced in {2} {3}, and we're using the {2} {4} SDK.", file, FileName, App.PlatformName, framework.Version, App.SdkVersion);
```
I know you're just moving code around but this one should be a warning (not just a log) since the code can fail at runtime (e.g. works locally, build on a bot with different Xcode, packaged product fails)

chamons replied:
Agree, but that would be everyone build with an older Xcode would have that warning. I'm assuming that is ok.

chamons commented:
I totally criss-crossed xcode and macOS versions in my head, you are right, a test should be straightforward.

monojenkins commented:
Build success

monojenkins commented:
Build success

rolfbjarne commented on tools/common/Assembly.cs:
```diff
@@ -328,6 +328,14 @@ public static LinkWithAttribute GetLinkWithAttribute (CustomAttribute attr)
+		void AddFramework (string file)
+		{
+			if (Driver.GetFrameworks (App).TryGetValue (file, out var framework) && framework.Version > App.SdkVersion)
+				ErrorHelper.Warning (135, "Did not link system framework '{0}' as Xcode SDK '{1}' is too old.", file, App.SdkVersion);
```
I didn't realize it when you asked in #mac, but this is now showing less information than before (in particular which assembly caused the warning, which can be quite useful). We also know which SDK version is required (so we can say that too), but this is probably less useful (although I always like to give as much information as possible in error/warning messages, since you never know what will be useful).

Also `Xcode SDK` is not entirely correct, if this is a watchOS project you'll end up with messages like `Did not link system framework 'XYZ' as Xcode SDK 4.0 is too old.`, and people end up confused because Xcode 4.0 is **old**.

chamons commented:
I added a test and updated the warning with less-terrible text (mostly stolen from Rolf's original log). Let me know if additional changes are needed.

monojenkins commented:
Build failure

VincentDondain commented on docs/website/mtouch-errors.md:
```diff
@@ -673,6 +673,15 @@ When using --recursive-directories, only 1 assembly should match
+
+They may fail at runtime as we are not linking against the required system framework.
+
+Consider updating to a newer Xcode, or if those APIs are not used, enabling [linking](https://docs.microsoft.com/en-us/xamarin/mac/deploy-test/linker).
```
So this 0135 error is a bit similar to https://docs.microsoft.com/en-us/xamarin/ios/troubleshooting/mtouch-errors#MT0091 in the sense you want people to update their Xcode or enable the linker. I think you should copy or take inspiration from the detailed "enable the linker" explanation in 0091.

Also I have to admit that I'm a bit confused with the general wording of this error, not sure I understand.

If I go line by line:

> Did not link system framework '{0}' as Xcode SDK '{1}' is too old.

ok, couldn't, link Xcode too old.

> Xamarin.iOS contains API bindings for the listed framework which is newer than the Xcode used.

The framework is newer than the Xcode used? That sounds weird. I don't understand this.

> They may fail at runtime as we are not linking against the required system framework.

The listed framework may fail or the API bindings :P ? The error is about a system framework and it's also the last element of the previous sentence so I think that grammatically "they" applies to the listed frameworks (am I wrong?).

Overall I'm confused with those 2 middle sentences that are supposed to explain the core of the issue (maybe it's just me).

If I try really hard to parse this error I get that it's basically https://docs.microsoft.com/en-us/xamarin/ios/troubleshooting/mtouch-errors#MT0091.
We can't "link" some frameworks because they are simply not available in the currently selected Xcode (too old). Some APIs in Xamarin.iOS depend on those frameworks (hence the error) and because we couldn't link them it's gonna fail at runtime. You can either upgrade Xcode to get the new framework or enable the linker so those APIs are skipped.

Anyway, we _might_ need to rephrase this, it's not crystal clear for me on the first read, it's very technical and hard to digest and I don't know if users will just understand this.

chamons commented:
@VincentDondain @rolfbjarne @spouliot - Redid the error message again, this time as a group effort.

monojenkins commented:
Build failure

chamons commented:
@bradumbaugh as well

bradumbaugh commented on docs/website/mtouch-errors.md:
```diff
@@ -675,12 +675,13 @@ When using --recursive-directories, only 1 assembly should match
-Consider updating to a newer Xcode, or if those APIs are not used, enabling [linking](https://docs.microsoft.com/en-us/xamarin/mac/deploy-test/linker).
+Alternatively, enable the managed [linker](https://docs.microsoft.com/en-us/xamarin/ios/deploy-test/linker?tabs=vsmac) to remove unused APIs, including (in most cases) the new ones which require the specified library. However, this will not work if your project requires APIs introduced in a newer SDK than the one your Xcode provides.
 
+As a last straw solution, use an older version of Xamarin.iOS that does not require these new SDKs to be present during the build process.
```
@chamons  "last-straw" needs a hyphen

bradumbaugh commented on docs/website/mmp-errors.md:
```diff
@@ -148,12 +148,13 @@ Consider updating your application and any dependencies to 64-bit.
-Consider updating to a newer Xcode, or if those APIs are not used, enabling [linking](https://docs.microsoft.com/en-us/xamarin/mac/deploy-test/linker).
+Alternatively, enable the managed [linker](https://docs.microsoft.com/xamarin/mac/deploy-test/linker) to remove unused APIs, including (in most cases) the new ones which require the specified library. However, this will not work if your project requires APIs introduced in a newer SDK than the one your Xcode provides.
 
+As a last straw solution, use an older version of Xamarin.Mac that does not require these new SDKs to be present during the build process.
```
@chamons "last-straw" needs a hyphen

VincentDondain commented on docs/website/mtouch-errors.md:
```diff
@@ -673,6 +673,16 @@ When using --recursive-directories, only 1 assembly should match
+
+The recommended way to fix this error is to upgrade Xcode to get the needed SDK. If you have multiple versions of Xcode installed or want to use an Xcode in a non-default location, make sure to set the correct Xcode location in your IDE's preferences.
+
+Alternatively, enable the managed [linker](https://docs.microsoft.com/en-us/xamarin/ios/deploy-test/linker?tabs=vsmac) to remove unused APIs, including (in most cases) the new ones which require the specified library. However, this will not work if your project requires APIs introduced in a newer SDK than the one your Xcode provides.
```
You can probably remove `?tabs=vsmac`, not sure it's a big deal though.

VincentDondain replied:
Idk if maybe there's a cookie that remembers which IDE the user chose and if that override their default.

monojenkins commented:
Build success
